src/lsp/symbols.rs

```diff
@@ -4,45 +4,84 @@ use crate::{
     quickfix_list::Location,
 };
 use lsp_types::{DocumentSymbolResponse, SymbolKind};
-use shared::icons::get_icon_config;
+use shared::{canonicalized_path::CanonicalizedPath, icons::get_icon_config};
 
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub(crate) struct Symbols {
     pub(crate) symbols: Vec<Symbol>,
 }
 
-impl TryFrom<DocumentSymbolResponse> for Symbols {
-    type Error = anyhow::Error;
+fn collect_document_symbols(
+    document_symbol: &lsp_types::DocumentSymbol,
+    symbols: &mut Vec<Symbol>,
+    parent_name: Option<String>,
+    path: &CanonicalizedPath,
+) -> Result<(), anyhow::Error> {
```

Comment:
I would prefer this function to return a vector instead of taking a mutable vector parameter, as that would be more composable and would result in fewer changes in this PR.